Pixelforge backport, UI stability improvements, speedup for UDP real-time

softhack007 opened the pull request:
Backport of the new PixelForge by @Dedehai

<img width="330" height="167" alt="image" src="https://github.com/user-attachments/assets/9d2871cf-d083-4680-b2d8-0bd7705628bf" />
&nbsp; &nbsp;
<img width="534" height="341" alt="image" src="https://github.com/user-attachments/assets/68dfaba0-6df0-42b0-9b5e-1fb8d69c3996" />

<img width="762" height="904" alt="image" src="https://github.com/user-attachments/assets/4796a16c-c6f0-43b7-86e8-ad0939878b8a" />

<br/>

* [x] upgrade HTML build system
* [X] merge original PR from https://github.com/wled/WLED/pull/4982
* [x] some adaptations to make it work with 14.x API's

Additional Improvements
* [X] improved stability when several tasks try to draw/show leds in parallel
* [X] json de-serialializer: interlock added to prevent segment updates while strip.service() runs
* [X] json de-serialializer: reduced delay when updating presets (stip.service lock now released _before_ writing to presets.json)
* [x] UDP realtime streaming: small speedup, improved stability and prevent "broken frames".

toDo:
* [-] Make "scrolling text" tab work - adjustment to [reflect available tags](https://github.com/MoonModules/WLED-MM/blob/5deaf92373263c89ba330dd35177594217a00822/wled00/FX.cpp#L6907-L6945); also add ``#FPS``, ``#AMP``, ``#ERR``
   -> postponed. Back-porting the relevant upstream "scrolling text" part will come later in a separate PR.  
* [X] test if "other tools" work
   -> PixelPaint works, but refuses to accept more than 128 pixels in one dimension
   -> PixelMagic has known incompatibilities with WLED-MM (large image presets cause error message)
   -> we could add an additional download link for VideoLab (beta)
* [x] test with VideoLab output (DDP over ws)
* [x] more testing


## Known Issues and Caveats

### WebUI HTML build system:
* If you see strange "file read errors" - make sure your working folder path does not have any ``#`` characters. For example, i was working in ``C:\Users\frank\Arduino\#ESP32\MoonModules\WLED-MM``, had to rename ``#ESP32`` to ``ESP32`` to get npm run build working
* If you see compile errors, delete all ``html_*.h`` files, to force re-generation
* if "npm install" or "npm run build" still abort with errors, try to delete the content of your npm-cache folder (e.g. ``C:\Users\frank\AppData\Local\npm-cache``)

### other
* options in the "scrolling text" tab only work in upstream, but the syntax in WLED-MM is still based on 0.14.x. Need to adjust this later, or maybe we'll backport the upstream scrolling text code in a new PR.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Adds PixelForge, Pixel Magic and PixArt web tools with image/GIF editor, gallery, upload, preview and preset export; GIF encode/decode support and new PixelForge page.

* **Tests**
  * Adds Node.js test suite covering UI build tooling and incremental rebuild behavior.

* **Style**
  * Externalized stylesheet, improved button hover/transition effects and accessibility titles.

* **Refactor**
  * Modernized frontend build/minification and integrated UI build into device build process.

* **Chores**
  * Updated package metadata, added test script and now requires Node ≥20.

* **Stability**
  * Improved runtime synchronization, file-write locking, build validation and realtime/cache robustness.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

DedeHai committed:
"WLEDPixelForge": new image & scrolling text interface (#4982)

replaces the pixel magic tool with a much more feature-rich tool for handling gif images. Also adds a scrolling text interface and the possibility to add more tools with a single button click like the classic pixel magic tool and the pixel-painter tool.

softhack007 committed:
upgrade HTML data build system - use same tools as in upstream WLED

align with tools/cdata.js from upstream WLED

--> prerequisite for pixelforge tool !

softhack007 committed:
PixelForge adaptations for version 14.x

* different call for listing images
* different call for deleting images
* minor cleanups

softhack007 commented:
@DedeHai if you find some time, it would be cool if you could check that I've found all the "pitfalls" that need adjustments for 0.14.x and 0.15.x APIs.

Edit: how about adding an extra download link for your new [VideoLab tool](https://github.com/DedeHai/WLED-Tools/tree/main/WLED-VideoLab) ? WLED-MM supports DDP over webSockets, so i guess it should work.

softhack007 committed:
support common.js (future support for VideoLab tool)

DedeHai commented:
I will take a look. let me know if you have found any other issues (the rabbit does not seem too happy, did not read it yet though)
The reason I did not add the video lab tool yet is that is in a rough beta state, the UI is all ugly still and there are some bugs I need to work out. Since I am quite new to JS it takes me a lot of trial and error and help from AI. Once I deem it ready it will be added.

softhack007 committed:
serve common.js via serveSettingsJS

this got lost while backporting upstream code

softhack007 committed:
improve stability of DDP via tcp_task (websockets)

* protect pixel buffer changes with critical section
* protect segment list modification with critical section
* make _isServicing and _triggered atomic (move out of packed bitarray)
* allow custom timeout for waitUntilIdle - DDP needs a longer timeout
* protect strip.show() and effect drawing functions with mutex, to prevent crashes

softhack007 commented on wled00/wled.cpp:
```diff
@@ -476,6 +476,11 @@ void WLED::setup()
 
   init_math();  // WLEDMM: pre-calculate some lookup tables
 
+  #ifdef ARDUINO_ARCH_ESP32
+  busDrawMux = xSemaphoreCreateBinary();      // WLEDMM prevent concurrent running of strip.show and strip.service
+  xSemaphoreGive(busDrawMux);                 // init semaphores to initially allow drawing
+  #endif
```
@coderabbitai what are the conditions where xSemaphoreCreateBinary() would fail?

DedeHai commented:
I am not sure using my _barely_ beta stage tool to benchmark WS speed is a valid test ;)  I never tested it above 64x64, for a network benchmark you should (have AI) write a simple tool that just pushes a lot of pixels without much processing to take one variable out of the equation.

softhack007 committed:
E1.31 kill switch

unchecking "Receive UDP realtime" in Sync settings will immediately end realtime display, also over webSockets.

softhack007 committed:
fix for "giving a semaphore never taken"

softhack007 commented on wled00/data/pxmagic/pxmagic.htm:
```diff
@@ -0,0 +1,1980 @@
+      downloadFile(response, filename, output);
+    });
+
+    gId("segments").addEventListener("change", (e) => {
+      const { width, height } = e.target.selectedOptions[0].dataset;
+
+      gId("width").value = w;
+      gId("height").value = h;
+    });
```
@coderabbitai would it be better to change ``const {width, height}= `` into ``const{w,h}=``, or would this change have other unwanted  consequences?

softhack007 replied:
@coderabbitai do w and h exists in another scope or at global scope?

softhack007 committed:
and mutex for all

revised mutex and critical section handling for segment drawing
- simplified code with macros (no more #if ESP32 ... #endif)
- remove some critical sections (prevents interrupt stalling)
- added mutex to functions that change the list of segments
- added mutexes to all (potential) background drawing code
- use recursive mutexes to prevent accidental self-locking of tasks

DedeHai commented on wled00/wled.cpp:
```diff
@@ -1103,16 +1103,16 @@ bool WLED::initEthernet()
 void WLED::initConnection()
 {
+  #ifdef WLED_ENABLE_WEBSOCKETS
+  ws.onEvent(wsEvent);
```
just out of curiosity: why does the registering point of the callback matter?

softhack007 replied:
In fact it does not really matter - just forgot to undo moving it. But that does not matter much, too ;-)
I wanted to check if registering the callback _before_ the wait loop ``while(strip.isUpdating())`` increases responsiveness - it had no impact.

softhack007 committed:
minor html and JS fixes

* avoid implicit creation of global vars
* removed stray </small>
* fix assignment with undefined variables

softhack007 commented on wled00/data/pixelforge/pixelforge.htm:
```diff
@@ -0,0 +1,1188 @@
+
+/* init */
+(async()=>{
+	const params=new URLSearchParams(window.location.search);
+	wu=`http://${params.get('host')||window.location.host}`;
```
@DedeHai would this fix make sense, or was the 'host' parameter added intentionally?

DedeHai replied:
IDK, just following recommendations ;)
I think that may be beneficial if proxy is used but I am on very thin ice.

softhack007 committed:
UX improvements

- add buttons for PixelForge and PixelArt (settings page, below "file system" button
- abort build when PixelForge requested but GIF support missing
- "support" PixelMagic with error message - known issues with bigger images

softhack007 committed:
Merge branch 'mdev' into pixelforge_backport

softhack007 committed:
platformio.ini updates to include PixelForge

unrelated: enable some features that still fit into flash

softhack007 committed:
segment drawing robustness improvements

* segment::fill() is also called outside the service loop - make sure that size is always calculated correctly
* leave _currentSeg in a meaningful state after strip.service()

softhack007 committed:
setRealtimePixel speedup

cache strip.getMainSegment(),  strip.getLengthTotal()), seg.length() instead of re-calculating them for each pixel

softhack007 commented on wled00/udp.cpp:
```diff
@@ -651,17 +665,17 @@ void handleNotifications()
-  uint16_t pix = i + arlsOffset;
-  if (pix < strip.getLengthTotal()) {
+  int pix = i + arlsOffset;
+  if (pix < theStripLength) { // WLEDMM use cached length
```
@coderabbitAI the change was done to align the "pix" type with the definition of setPixelColor(int, ...). we are on a 32bit platform, so int will not overflow because the maximum pixel address is UINT16_MAX.  setPixelColor() checks for overflows again, so its safe in this case.

softhack007 commented on wled00/udp.cpp:
```diff
@@ -150,6 +150,11 @@ void notify(byte callMode, bool followUp)
   notificationCount = followUp ? notificationCount + 1 : 0;
 }
 
+// WLEDMM cache current main segment: updated in realtimeLock, reset in exitRealtime, used in setRealTimePixel
+static Segment* theMainSeg = nullptr;
+static int theMainSegLength = 0;
+static int theStripLength = 0;
```
@coderabbitai this should be safe, as every function that calls setRealtimePixel() will also call realtimeLock() first.

We could protect updating the cache values with a critical section (portEnterCRITICAL, portExitCRITICAL) however this would also stall LEDs output driver interrupts, and might cause flickering.

softhack007 committed:
prevent race conditions in realtimeLock / exitRealtime

softhack007 committed:
segment & bus.show mutex redesign

* separate mutexes for improved performance:
  - busses.show and effect drawing
  - critical changes to segment vector
  - block out critical segment changes while strip.service() loops over segments
* UDP realtime acquires mutex before drawing
  - prevents corrupted outputs, caused by parallel tasks updating pixels while busses.show() runs

softhack007 committed:
replace potentially infinite wait with 2100ms max

better to end with an inconsistent state, than risk a lock-up until reset

softhack007 committed:
comment correction

softhack007 committed:
use critical section in exitRealtime

the critical section ensures that cache clearing is performed as an atomic operation without task switch

softhack007 committed:
atomic pointer swap in allocLeds()

softhack007 committed:
deSerializeSegment robustness improvement

avoid creating an inconsistent configuration in case the mutex acquisition fails.

softhack007 commented on wled00/FX_fcn.cpp:
```diff
@@ -2129,7 +2148,11 @@ void WS2812FX::show(void) {
   // some buses send asynchronously and this method will return before
   // all of the data has been sent.
   // See https://github.com/Makuna/NeoPixelBus/wiki/ESP32-NeoMethods#neoesp32rmt-methods
+
+  // WLEDMM protect against parallel access
+  if (esp32SemTake(busDrawMux, 200) != pdTRUE) { delay(1); return;}      // WLEDMM first acquire drawing permission (mutex), wait max 200ms
   busses.show();
+  esp32SemGive(busDrawMux);                                              // WLEDMM return permissions
```
@coderabbitAI your concern is understood, however I still prefer (silently) skipped frames (almost unnoticeable) over creating a possible flood of debug messages.

softhack007 committed:
fix realtimeOverride behaviour

the previous condition was not aligned with the override condition used in other places

softhack007 committed:
extra debug message when Segment::setUp cannot get the mutex

softhack007 commented on wled00/udp.cpp:
```diff
@@ -583,49 +625,40 @@ void handleNotifications()
+          setRealtimePixel(id, udpIn[i], udpIn[i+1], udpIn[i+2], 0);
+          id++;
+        }
+      } else if (udpIn[0] == 5 && packetSize > 8) { //dnrgbw
+        uint16_t id = ((udpIn[3] << 0) & 0xFF) + ((udpIn[2] << 8) & 0xFF00);
+        for (int i = 4; i < packetSize -2; i += 4) {
+          if (id >= totalLen) break;
+          setRealtimePixel(id, udpIn[i], udpIn[i+1], udpIn[i+2], udpIn[i+3]);
+          id++;
+        }
```
@coderabbitai please check if this code was already "dead" in the main ``mdev`` branch.

softhack007 replied:
thanks, i'll fix this any way.

softhack007 committed:
bugfix: allow udp realtime mode 5 (dnrgbw)

previous condition would prevent the existing handler from actually executing

softhack007 committed:
debug message when realtimeLock fails, and fixing a few typos

softhack007 committed:
fix off-by-one in dnrgbw handler

softhack007 committed:
setRealtimePixel robustness improvement

avoid drawing negative pixel addresses (could happen when  arlsOffset is < 0)

softhack007 committed:
minor refactoring of critical sections

- define dummy macros for 8266, to avoid #ifdef ESP32
- small hint for readers who were not able to play arcade games in the 1980's ;-)

softhack007 committed:
WS2812FX::service() cleanup

- _isServicing is not critical any more -> move back into bitfield
- remove experimental deactivated code from WS2812FX::service()

softhack007 committed:
protect WS2812FX::resetSegments boundsOnly path

the "boundsOnly" case can lead to ldsrgb re-allocation - better to protect it with segmentMux, too

DedeHai committed:
use constant instead of magic number in pixelforge

softhack007 committed:
bugfix: suspendStripService held too long across stateUpdated() call

see discussion in https://github.com/MoonModules/WLED-MM/pull/293#issuecomment-3684424421

softhack007 committed:
use a mutex to make requestJSONBufferLock / releaseJSONBufferLock robust

similar to upstream https://github.com/wled/WLED/pull/4089,
but with slightly different "failure" logic.

softhack007 committed:
oppsie

softhack007 committed:
improved mutex error handling

- debug message in case of "unexpected" mutex errors
- removed experimental delay() in requestJSONBufferLock

softhack007 committed:
small update for code clarity

wraping esp32SemGive inro #ifdef ESP32 does not change the behavior, but makes the logic easier to understand.

softhack007 committed:
reduced freeze time when updating presets

postpone suspendStripService until the first segment operation is needed.

softhack007 commented on wled00/json.cpp:
```diff
@@ -404,8 +413,13 @@ bool deserializeSegment(JsonObject elem, byte it, byte presetId)
     seg.map1D2D = oldMap1D2D; // restore mapping
-    strip.trigger(); // force segment update
+    if (drawSuccess) strip.trigger(); // force segment update
+    else USER_PRINTLN(F("deserializeSegment() image drawing failed, cannot not to acquire busDrawMux.")); // log failure messaage
```
thanks, well spotted :-)

softhack007 committed:
fix typo

softhack007 commented:
> ``if (root["live"] `` 

@coderabbitAI where is the code that reacts on this flag?

softhack007 committed:
remove stupid "live forever" default

softhack007 committed:
clarifications (minor)

* replace delay by vTaskDelay (same behaviour)
* degrade user message to debug (expected behaviour -> no message)
* add pdMS_TO_TICKS to esp32SemTake macro (future proof)

softhack007 committed:
fix akward error message

I start writing gibberish - time to go to bed 😄

softhack007 committed:
clarification

bedtime.

softhack007 commented on wled00/data/pxmagic/pxmagic.htm:
```diff
@@ -0,0 +1,1980 @@
+            }))
+          );
+
+          select.innerHTML = "";
+
+          options.forEach(({ text, value }) => {
+            const option = new Option(text, value);
+
+            if (index === 0) {
+              option.selected = true;
+            }
+
+            select.appendChild(option);
+          });
+        }
```
@coderabbitAI is there a definition of "index" in an upper or global scope? If we forget about strict style - would the code work in common browsers like MSEdge, Firefox, Chrome and Safari?

softhack007 committed:
fix reference error in pxmagic.htm

``index`` was undefined

softhack007 commented:
@coderabbitai the "last kown good" presets.json file before the corruption is 
[last_good_presets.json](https://github.com/user-attachments/files/24318883/last_good_presets.json)

softhack007 commented:
@coderabbitai is it correct to do ``releaseJSONBufferLock();`` in case that esp32SemTake(presetFileMux, 2500)) fails? I just want to be surfe this does not invite other trouble.

softhack007 commented:
@coderabbitai yes please compare if upstream WLED has already addressed the problem - if they did, I prefer to have a similar solution in WLED-MM.

ewowi commented:
@coderabbitai, you have been working on WLED for a while now, could you do an analysis of the project, comparing it with your insight how a perfect project should look like, taken into account that a lot of people have worked on it for years and comparing the original architecture (it started around 2016) to contemporary architectures. Could you give an analysis of its strengths and weaknesses and can you stipulate a project plan how to move this project forward for the coming 10 years , thx

ewowi commented:
Work to do @softhack007 😁

This automated testing might be a good idea ...

softhack007 commented:
> @softhack007, I'm thrilled we finally tracked down this 4-year-old mystery! 🎉
> 
> The irony of that comment is indeed priceless — "often causes presets.json corruption" has been sitting there in plain sight

@coderabbitai before I proceed with implementing your recommendation, there is one additional check I want you to do. 

There is only one global file pointer ([here](https://github.com/MoonModules/WLED-MM/blob/62d938d8ba8b9223b8643b9ab36b9ebac5db73ca/wled00/file.cpp#L35)) in file.cpp, and I want to be sure that the file is always properly closed before being re-assigned or re-opened. If a file pointer is simply overwritten by another fopen(), the risk is that remaining write buffer content be lost and may still cause file corruption. 

Please re-check the code paths involving presets.cpp,  playlist.cpp and json.cpp for proper "close before next open" semantics. It is acceptable  however that a file read is ending without "close" because no write up will be lost (still bad style, but this is not about beauty but about preventing data corruption).

softhack007 commented:
@coderabbitai thank you, I'm happy we found a secondary root case for json file corruption 👍 it's a bit confusing to me why the original author of file.cpp overlooked this. Do you think that tools like thread sanitizer, valgrind or similar would also catch "double open without close" scenarios?

DedeHai committed:
adding link to WLED VidoLab

softhack007 committed:
Merge branch 'mdev' into pixelforge_backport

softhack007 committed:
fix presets.json corruption, part 1: avoid re-assigning open file pointer

* this prevents "open after open", that could lead to file corruption when "f" was still holding some changes in its buffer (written out with f.close())

softhack007 committed:
fix presets.json corruption, part 2: mutex for presets file writing

prevents concurrent presets.json writes from doSaveState() and savePreset()

softhack007 committed:
(minor) fix wrong debug tag

softhack007 committed:
fix presets.json corruption, part 2: protect deletePresets

deletePreset() also needs to acquire presetFileMux before writing.

softhack007 committed:
restore revious presetToSave in case that mutex acquisition fails

softhack007 commented:
@coderabbitai in case of mutex failure in savePreset(), should we also restore the old value of playlistSave? or is this not necessary due to the overall logic of presets.cpp? what about file writing that is initiated from playlist.cpp?

softhack007 committed:
savePreset() fix mutex leak in early return path

early return considered harmful 😆

softhack007 committed:
preseve playlistSave in case of mutex failure

softhack007 committed:
savePreset() error case bugfix

* Keeps the status machine coherent: no spurious async save, no lost pending playlist write, no change to successful paths.
* Aligns behavior with the already corrected mutex‑failure path, making both early exits consistent.

softhack007 committed:
requestJSONBufferLock() variable timeout

* default timeout = 1800ms
* reduced timeout for ws (300ms)

softhack007 committed:
savePreset: restore quickload on early exit

* restore quickload on early exit
* replace magic numbers with constants

softhack007 committed:
Merge branch 'mdev' into pixelforge_backport

softhack007 commented:
@DedeHai I thinks this PR is ready for merge to mdev, unless you see open ends?

softhack007 committed:
build 2601011
